### ISSUE TITLE
Issues in running harbor with restricted psp

### ISSUE THREAD

**Issue by tvvignesh** [goharbor/harbor-helm#756]
Hi. This is how I have setup psp in my cluster: https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/policy/restricted-psp.yaml (restricted PSP which disables root access, previlege escalation and more as recommended everywhere)

I am not able to run harbor with this policy set (have tried an entire day).

1) The chart does not have any securityContext and so it fails.

2) Postgres init containers wont work since it runs as user 0 which is denied as per this policy and disabling initcontainers will lead to permission denied for `/var/lib/postgresql/data`

3) I even tried chart from bitnami and even that fails since that also tries to run the initcontainer with user 0 which is not allowed.

This happens for almost all deployments within harbor chart - tried modifying the chart manually but they are leading to different errors.

Can someone please help? I am trying to set harbor up in a psp restricted cluster but am not able to. If anyone has already done it, can I have some reference? Thanks.

CC: @ywk253100 @reasonerjt

**Comment by msiebuhr** [comment]
As to (2), Postgres really _really_ wants to own the folder it's storing data in. Since it is not running as root, it cannot change permissions on the mount (owned by root:postgres, if you set `securityContext -> fsGroup: 999). If you set $PGDATA=/var/lib/postgresql/data/pgdata`, it will create the required sub-folder and thus be the owner of it.

Our minimal deployment made like this:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: postgres
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: postgres
    spec:
      securityContext:
        runAsUser: 999 # See https://github.com/docker-library/postgres/blob/bfc5d81c8f5647c690f452dc558e64fddb1802f6/10/Dockerfile
        runAsGroup: 999
        fsGroup: 999
      containers:
        - name: postgres
          image: postgres:10.4
          imagePullPolicy: "IfNotPresent"
          ports:
            - containerPort: 5432
          envFrom:
            - configMapRef:
                name: postgres-config
          env:
            - name: PGDATA
              value: /var/lib/postgresql/data/pgdata
          volumeMounts:
            - mountPath: /var/lib/postgresql/data
              name: postgredb
      volumes:
        - name: postgredb
          persistentVolumeClaim:
            claimName: postgres-pv-claim
```

**Comment by msiebuhr** [comment]
On a related note, we've run into a lack of well-defined users. From starting up *notary*: 

    Error: container has runAsNonRoot and image has non-numeric user (notary), cannot verify user is non-root

It turns out that the Dockerfile wants to run as user 'notary', but doesn't define it (see https://hub.docker.com/layers/goharbor/notary-server-photon/v2.1.0/images/sha256-28ff22aef3f3b75244c531640b0a9c147f589d7ee362ea1dcf82c462161641f3?context=explore). I haven't been able to find the upstream Dockerfile to submit a change-request...

Alternatively, we the helm-chart really should set a `securityContext` with some appropriate values. Running needlessly as root can only be a win, IMHO.

**Comment by msiebuhr** [comment]
Here's the patch I worked up to get it running on our setup:

```diff --git a/templates/chartmuseum/chartmuseum-dpl.yaml b/templates/chartmuseum/chartmuseum-dpl.yaml
index dc0157f..6670667 100644
--- a/templates/chartmuseum/chartmuseum-dpl.yaml
+++ b/templates/chartmuseum/chartmuseum-dpl.yaml
@@ -36,6 +36,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.chartmuseum.serviceAccountName }}
       serviceAccountName: {{ .Values.chartmuseum.serviceAccountName }}
diff --git a/templates/clair/clair-dpl.yaml b/templates/clair/clair-dpl.yaml
index 35005e8..e3d8d42 100644
--- a/templates/clair/clair-dpl.yaml
+++ b/templates/clair/clair-dpl.yaml
@@ -29,6 +29,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.clair.serviceAccountName }}
       serviceAccountName: {{ .Values.clair.serviceAccountName }}
diff --git a/templates/core/core-dpl.yaml b/templates/core/core-dpl.yaml
index 10002f1..3ce2193 100644
--- a/templates/core/core-dpl.yaml
+++ b/templates/core/core-dpl.yaml
@@ -30,6 +30,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.core.serviceAccountName }}
       serviceAccountName: {{ .Values.core.serviceAccountName }}
diff --git a/templates/jobservice/jobservice-dpl.yaml b/templates/jobservice/jobservice-dpl.yaml
index cafc6fb..87b2c78 100644
--- a/templates/jobservice/jobservice-dpl.yaml
+++ b/templates/jobservice/jobservice-dpl.yaml
@@ -36,6 +36,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.jobservice.serviceAccountName }}
       serviceAccountName: {{ .Values.jobservice.serviceAccountName }}
diff --git a/templates/notary/notary-server.yaml b/templates/notary/notary-server.yaml
index 6cb8023..4f75d7f 100644
--- a/templates/notary/notary-server.yaml
+++ b/templates/notary/notary-server.yaml
@@ -25,6 +25,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.notary.server.serviceAccountName }}
       serviceAccountName: {{ .Values.notary.server.serviceAccountName }}
diff --git a/templates/notary/notary-signer.yaml b/templates/notary/notary-signer.yaml
index f4ee98e..ea6ca9a 100644
--- a/templates/notary/notary-signer.yaml
+++ b/templates/notary/notary-signer.yaml
@@ -21,6 +21,8 @@ spec:
         checksum/secret: {{ include (print $.Template.BasePath "/notary/notary-secret.yaml") . | sha256sum }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.notary.signer.serviceAccountName }}
       serviceAccountName: {{ .Values.notary.signer.serviceAccountName }}
diff --git a/templates/portal/deployment.yaml b/templates/portal/deployment.yaml
index 855830f..c74e5fb 100644
--- a/templates/portal/deployment.yaml
+++ b/templates/portal/deployment.yaml
@@ -26,6 +26,9 @@ spec:
 {{ toYaml .Values.portal.podAnnotations | indent 8 }}
 {{- end }}
     spec:
+      securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
       {{- with .Values.imagePullSecrets }}
       imagePullSecrets:
         {{- toYaml . | nindent 8 }}
diff --git a/templates/redis/statefulset.yaml b/templates/redis/statefulset.yaml
index be710af..7c38dcc 100644
--- a/templates/redis/statefulset.yaml
+++ b/templates/redis/statefulset.yaml
@@ -25,6 +25,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 999
+        runAsGroup: 999
         fsGroup: 999
 {{- if .Values.redis.internal.serviceAccountName }}
       serviceAccountName: {{ .Values.redis.internal.serviceAccountName }}
diff --git a/templates/registry/registry-dpl.yaml b/templates/registry/registry-dpl.yaml
index 24b6b60..b181da4 100644
--- a/templates/registry/registry-dpl.yaml
+++ b/templates/registry/registry-dpl.yaml
@@ -36,6 +36,8 @@ spec:
 {{- end }}
     spec:
       securityContext:
+        runAsUser: 10000
+        runAsGroup: 10000
         fsGroup: 10000
 {{- if .Values.registry.serviceAccountName }}
       serviceAccountName: {{ .Values.registry.serviceAccountName }}
```

(I still have an issue of notary-signer not being able to connect to the external database, but I suspect that has other causes...)

**Comment by tvvignesh** [comment]
I will probably wait for some more time till an official patch comes from harbor else I will have to be going back and forth 😅 Restricted PSP is really hard to work with since a lot of tools are having issues because of it 😢

**Comment by ywk253100** [comment]
@msiebuhr Thanks for the suggested solution.
One more thing we should consider is that how can we handle the upgrade for database as we change the directory where the data stores?

**Comment by msiebuhr** [comment]
@ywk253100 - true, there is some migration to be done.

Given that everyone that hasn't run into this issue _is_ running as root, we could do an init-container for postgres that ensures the database has been moved to a sub-folder. Then the upgrade to PSP-enabled Harbor can happen in a subsequent upgrade.

**Comment by jsuchome** [comment]
@tvvignesh what Kubernetes version are you running? I'm trying Harbor chart 1.5.0 with the same restricted policy as you provided, but it does not complain about root user running that init container for database..

**Comment by tvvignesh** [comment]
@jsuchome Hi. Currently running 1.18.9-gke.1501 (in GKE). Wondering why though (haven't changed anything much different from the harbor chart) - I was running `v1.4.2` of the chart though (pointing to Harbor `2.x`). Havent tried the new charts since thios was scheduled to be fixed on `v1.6` as per @ywk253100 

Btw, practically it should not work for you since as per this policy here: https://github.com/goharbor/harbor-helm/blob/1b676d68401b1c15cdb03a4e34787b2ed24756b4/templates/database/database-ss.yaml#L38 it uses `runAsUser: 0` which is root and is prohibited as per the psp I linked.

**Comment by jsuchome** [comment]
Thanks. Yes, I understand that it should fail for me because of init container running as root. So I was surprised to see it actually did _not_ fail in my cluster. I just spinned 1.8.10 cluster in Azure and it failed as expected.

I could be wrong, but I suspect the Kubernetes version.
